app/vmselect/vmui.go

```diff
@@ -0,0 +1,148 @@
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/fs"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
+)
+
+// more information how to use this flag please check this link
+// https://github.com/VictoriaMetrics/VictoriaMetrics/tree/master/app/vmui/packages/vmui/public/dashboards
+var (
+	vmuiCustomDashboardsPath = flag.String("vmui.customDashboardsPath", "", "Optional path to vmui predefined dashboards")
+)
+
+var predefinedDashboards = &dashboardsData{}
+
+// How frequently check defined vmuiCustomDashboardsPath
+const updateDashboardsTimeout = time.Second * 5
+
+// dashboardSetting represents dashboard settings file struct
+// fields of the dashboardSetting you can find by following next link
+// https://github.com/VictoriaMetrics/VictoriaMetrics/tree/master/app/vmui/packages/vmui/public/dashboards
+type dashboardSetting struct {
+	Title    string         `json:"title,omitempty"`
+	Filename string         `json:"filename,omitempty"`
+	Rows     []dashboardRow `json:"rows"`
+}
+
+// panelSettings represents fields which used to show graph
+// fields of the panelSettings you can find by following next link
+// https://github.com/VictoriaMetrics/VictoriaMetrics/tree/master/app/vmui/packages/vmui/public/dashboards
+type panelSettings struct {
+	Title       string   `json:"title,omitempty"`
+	Description string   `json:"description,omitempty"`
+	Unit        string   `json:"unit,omitempty"`
+	Expr        []string `json:"expr"`
+	Alias       []string `json:"alias,omitempty"`
+	ShowLegend  bool     `json:"showLegend,omitempty"`
+	Width       int      `json:"width,omitempty"`
+}
+
+// dashboardRow represents panels on dashboard
+// fields of the dashboardRow you can find by following next link
+// https://github.com/VictoriaMetrics/VictoriaMetrics/tree/master/app/vmui/packages/vmui/public/dashboards
+type dashboardRow struct {
+	Title  string          `json:"title,omitempty"`
+	Panels []panelSettings `json:"panels"`
+}
+
+// dashboardsData represents all dashboards settings
+type dashboardsData struct {
+	DashboardsSettings []dashboardSetting `json:"dashboardsSettings"`
+
+	mx                sync.Mutex
+	expireAtTimestamp time.Time
+}
+
+func handleVMUICustomDashboards(w http.ResponseWriter) {
+	path := *vmuiCustomDashboardsPath
+	if path == "" {
+		writeSuccessResponse(w, predefinedDashboards)
+		return
+	}
+
+	if !fs.IsPathExist(path) {
+		writeErrorResponse(w, fmt.Errorf("cannot find folder pointed by -vmui.customDashboardsPath=%q", path))
+		return
+	}
+
+	requestTime := time.Now()
+	if predefinedDashboards.expireAtTimestamp.Before(requestTime) {
+		settings, err := collectDashboardsSettings(path)
+		if err != nil {
+			writeErrorResponse(w, fmt.Errorf("cannot collect dashboards settings by -vmui.customDashboardsPath=%q", path))
+			return
+		}
+
+		predefinedDashboards.expireAtTimestamp = time.Now().Add(updateDashboardsTimeout)
```

Comment:
There is a data race here when multiple concurrently running goroutines update `expireAtTimestamp` field. There is also another data race when there are other goroutines which read the `expireAtTimestamp` field at the line 79 while the currently running goroutine updates this field here. This may result in invalid data read at the line 79.

Reply:
the cache was removed. If we need this cache in the future, I can implement a more reliable using goroutine